bin/storer.py

```diff
@@ -0,0 +1,102 @@
+        parser.add_argument('--bundles', default=self.DEFAULT_STAGED_BUNDLES_URL,
+                            help="S3 URL of staged bundles")
+        parser.add_argument('--bundle',
+                            help="S3 URL of a single staged bundle")
+        parser.add_argument('--smaller', type=int, default=1024*1024,  # 1 TB
+                            help="only store bundles with data files smaller than this many MB")
+        parser.add_argument('--dss-endpoint', default=DataStoreAPI.DEFAULT_DSS_URL,
+                            help="URL of DSS API endpoint")
+        parser.add_argument('-r', '--use-rest-api', action='store_true', default=False,
+                            help="use REST API directly, instead of python bindings")
+        parser.add_argument('-j', '--jobs', type=int, default=1,
+                            help="parallelize with this many jobs (implies --terse)")
+        parser.add_argument('-q', '--quiet', action='store_true', default=False,
+                            help="silence is golden")
+        parser.add_argument('-t', '--terse', action='store_true', default=False,
+                            help="terse output, one character per file")
+        parser.add_argument('-i', '--report-task-ids', action='store_true', default=False,
+                            help="when ")
+        parser.add_argument('-l', '--log', default=None,
+                            help="log verbose output to this file")
+        self.args = parser.parse_args()
+        if self.args.jobs > 1:
+            quiet = True
+            terse = True
+        else:
+            quiet = self.args.quiet
+            terse = self.args.terse
+        logger.configure(self.args.log, quiet=quiet, terse=terse)
+        self.storer_options = {'use_rest_api': self.args.use_rest_api, 'report_task_ids': self.args.report_task_ids}
+
+        if self.args.bundle:
+            self._store_bundle(parse_url(self.args.bundle))
+            print("\n")
+        else:
+            self._store_bundles(parse_url(self.args.bundles))
+
+    def _store_bundles(self, bundles_url: Url):
+        bundle_paths = StagedBundleFinder().paths_of_bundles_under(bundles_url)
+        if self.args.jobs > 1:
+            self.store_bundles_in_parallel(bundles_url, bundle_paths)
+        else:
+            for path in bundle_paths:
+                self._store_bundle(Url(scheme=bundles_url.scheme, host=bundles_url.host, path=path.lstrip('/')))
+
+    def store_bundles_in_parallel(self, bundles_url: Url, bundle_paths: list):
+        global executor
+        signal.signal(signal.SIGINT, self.signal_handler)
+        executor = ProcessPoolExecutor(max_workers=self.args.jobs)
+        bundle_number = 0
+        for path in bundle_paths:
+            bundle_number += 1
+            executor.submit(self._store_bundle,
+                            Url(scheme=bundles_url.scheme, host=bundles_url.host, path=path.lstrip('/')))
+        executor.shutdown()
```

Comment:
Concurrent.futures is often bad at handling Ctrl-C. Can you check that it works OK? (I haven't run this code myself yet)

Reply:
I've been working with this a fair amount now and it seems reliable after I made the following change: as soon as you have forked you need to reset SIGINT back to SIG_DFL, otherwise every process tries to initiate the shutdown procedure.